javascript.md

```diff
@@ -0,0 +1,793 @@
+
+Whenever you have to manipulate an array-like object, cast it to an array.
+
+##### Bad
+
+```js
+var divs = document.querySelectorAll('div');
+
+for (i = 0; i < divs.length; i++) {
+  console.log(divs[i].innerHTML);
+}
+```
+
+##### Good
+
+```js
+var divs = document.querySelectorAll('div');
+
+[].slice.call(divs).forEach(function (div) {
+  console.log(div.innerHTML);
+});
+```
+
+However, be aware that there is a [substantial performance hit][22] in V8 environments when using this approach on `arguments`. If performance is a major concern, avoid casting `arguments` with `slice` and instead use a `for` loop.
+
+#### Bad
+```js
+var args = [].slice.call(arguments);
+```
+
+#### Good
+```js
+var i;
+var args = new Array(arguments.length);
+for (i = 0; i < args.length; i++) {
+    args[i] = arguments[i];
+}
+```
+
+Don't declare functions inside of loops.
+
+##### Bad
+
+```js
+var values = [1, 2, 3];
+var i;
+
+for (i = 0; i < values.length; i++) {
+  setTimeout(function () {
+    console.log(values[i]);
+  }, 1000 * i);
+}
+```
+
+```js
+var values = [1, 2, 3];
+var i;
+
+for (i = 0; i < values.length; i++) {
+  setTimeout(function (i) {
+    return function () {
+      console.log(values[i]);
+    };
+  }(i), 1000 * i);
+}
+```
+
+##### Good
+
+```js
+var values = [1, 2, 3];
+var i;
+
+for (i = 0; i < values.length; i++) {
+  setTimeout(function (i) {
+    console.log(values[i]);
+  }, 1000 * i, i);
+}
+```
+
+```js
+var values = [1, 2, 3];
+var i;
+
+for (i = 0; i < values.length; i++) {
+  wait(i);
+}
+
+function wait (i) {
+  setTimeout(function () {
+    console.log(values[i]);
+  }, 1000 * i);
+}
+```
+
+Or even better, just use `.forEach` which doesn't have the same caveats as declaring functions in `for` loops.
+
+##### Better
+
+```js
+[1, 2, 3].forEach(function (value, i) {
+  setTimeout(function () {
+    console.log(value);
+  }, 1000 * i);
+});
+```
+
+Whenever a method is non-trivial, make the effort to **use a named function declaration rather than an anonymous function**. This will make it easier to pinpoint the root cause of an exception when analyzing stack traces.
+
+##### Bad
+
+```js
+function once (fn) {
+  var ran = false;
+  return function () {
+    if (ran) { return };
+    ran = true;
+    fn.apply(this, arguments);
+  };
+}
+```
+
+##### Good
+
+```js
+function once (fn) {
+  var ran = false;
+  return function run () {
+    if (ran) { return };
+    ran = true;
+    fn.apply(this, arguments);
+  };
+}
+```
+
+Avoid keeping indentation levels from raising more than necessary by using guard clauses instead of flowing `if` statements.
+
+##### Bad
+
+```js
+if (car) {
+  if (black) {
+    if (turbine) {
+      return 'batman!';
+    }
+  }
+}
+```
+
+```js
+if (condition) {
+  // 10+ lines of code
+}
+```
+
+##### Good
+
+```js
+if (!car) {
+  return;
+}
+if (!black) {
+  return;
+}
+if (!turbine) {
+  return;
+}
+return 'batman!';
```

Comment:
Some styles guides would have problems with the inconsistent return values shown here http://eslint.org/docs/rules/consistent-return